rfcs/20200817-jit-code-generation-for-aarch64/README.md

```diff
@@ -0,0 +1,256 @@
+`Xbyak_aarch64` functions correspond to the translation result of x64 
+instruction. 
+
+```
+xbyak_translator_aarch64
+|-- gen
+|-- sample
+|-- test
+|-- xbyak
+|-- translator
+    |-- include
+    |   `-- xbyak_translator_for_aarch64
+    |       |-- instructions
+    |-- tests
+    |-- third_party
+    |   |-- xbyak_aarch64
+    `-- tools
+```
+
+Xbyak_translator_aarch64 has additional 'translator' directory, which contains 
+the files for the translator functions. 
+`translator/indluce/xbyak_translator_for_aarch64` directory contains files 
+that implement the main functions of the translation, such as 
+`decodeAndTransToAArch64` and functions called by `decodeAndTransToAArch64`. 
+The rules for translating x64 instructions to AArch64 ones are implemented 
+in files in `translator/indluce/xbyak_translator_for_aarch64/instructions` 
+directory. At the moment, `Xbyak_translator_aarch64` supports 200 
+instructions that are used when running oneDNN on a CPU which can execute 
+AVX512/AVX2/AVX/SSE4.1 instructions.
+
+### Adding a dependency on xed at the build stage:
+
+To build oneDNN with Xbyak_translator_aarch64, it must be assumed that 
+a pre-build xed library is available, built from the xed sources. 
+Additional CMake modules will be needed, such as `FindXED.cmake` and 
+`XED.cmake` which will only impact AArch64 builds. 
+[We modified just a few lines](https://github.com/fujitsu/xed/commit/92ee2830ca48fed76baa0a36e7d6ef8a3afe5741) 
+of the original xed sources and disclosed as 
+[fujitsu/xed](https://github.com/fujitsu/xed/tree/fjdev "fujitsu/xed"). 
+
+### Adding primitive entry to src/cpu/cpu_\*_list.cpp
+
+As JIT code for AArch64 is ready to be generated at this point, primitives are 
+added to `src/cpu/cpu_*_list.cpp` as follows not only for convolution but also 
+eltwise, batch_normalization, pooling, etc. This proposal is the same as the 
+proposal in 
+[#795](https://github.com/oneapi-src/oneDNN/pull/795) and the implementation 
+for x64.
+
+```
+static const std::map<conv_impl_key_t, std::vector<pd_create_f>> impl_list_map {
+    // FWD fp
+    {{forward, f32, f32, f32}, {
+        CPU_INSTANCE_X64(jit_avx512_common_dw_convolution_fwd_t)
+        CPU_INSTANCE_X64(jit_avx512_common_1x1_convolution_fwd_f32_t)
+        CPU_INSTANCE_X64(jit_avx512_core_f32_wino_conv_2x3_fwd_t)
+        CPU_INSTANCE_X64(jit_avx512_core_f32_wino_conv_4x3_fwd_t)
+        CPU_INSTANCE_X64(jit_avx512_common_convolution_winograd_fwd_t)
+        CPU_INSTANCE_X64(jit_avx512_common_convolution_fwd_t<f32>)
+        CPU_INSTANCE_X64(jit_avx2_dw_convolution_fwd_t)
+        CPU_INSTANCE_X64(jit_avx2_1x1_convolution_fwd_t)
+        CPU_INSTANCE_X64(jit_sse41_dw_convolution_fwd_t)
+        CPU_INSTANCE_X64(jit_sse41_1x1_convolution_fwd_t)
+        CPU_INSTANCE_X64(jit_avx2_convolution_fwd_t)
+        CPU_INSTANCE_X64(jit_sse41_convolution_fwd_t)
+        CPU_INSTANCE_AARCH64(jit_aarch64_sve_512_common_dw_convolution_fwd_t)         <- Added
```

Comment:
Could you please elaborate on the interplay with [ArmCL integration](https://github.com/oneapi-src/oneDNN/tree/rfcs/rfcs/20200731-Using-Arm-Compute-Library-kernels-on-AArch64#adding-a-cpu_instance_aarch64-macro)? How would the dispatching and priorities be set?

Reply:
Like this.
CPU_INSTANCE_AARCH64(jit_aarch64_sve_512)
CPU_INSTANCE_AARCH64(ArmCL)

I.e. use jitted code if jitted code supports your SIMED instruction, use ArmCL otherwise.

Reply:
To be precise, the code looks like this
```
CPU_INSTANCE_AARCH64(jit_aarch64_sve_512_common_dw_convolution_fwd_t)         <- Added
CPU_INSTANCE_AARCH64(arm_gemm_convolution_fwd_t)                              <- ArmCL
```

Reply:
Thanks! I think for `sve_512` it is fair to put it above ArmCL, as it is very dedicated kernel for Fujitsu A64FX hardware (not sure if ArmCL supports SVE of size 512 bit at all).
The more concerning part is SIMD&FP (what I referred as Neon before, probably inappropriately). Assuming both ArmCL and JIT `aarch64_simdfp` kernels could be run how would we know which one is better to dispatch?

Casting @nSircombe to express his opinion. I am fine with either :)

Reply:
At the moment, there is only A64FX that supports SVE instructions and jit_sve_512 is tunned to A64FX so that I'd like to set jit_sve_512 ahead of ArmCL. As for SIMD&FP, I think it's better to measure jit_simdfp and ArmCL on several CPUs that support SIMD&FP, for example A64FX, ThunderX2, etc, and prioritize based on the results. In the future, we may need a way to switch between JIT and ArmCL in a flexible way.

Reply:
Just a quick note that our PR #820 introduces addition macros, e.g. `CPU_INSTANCE_ARCH64_ACL`, which are only enabled when oneDNN is built with `-DDNNL_AARCH64_USE_ACL=ON` this is to ensure that AArch64 builds _without_ Compute Library on AArch64 will still function. With Compute Library presently a dependency, and not part of the oneDNN build, this seams prudent - in the longer term, closer Compute Library integration into the oneDNN build could allow this to be removed.

Reply:
Certainly the suggested ordering...
```
...
CPU_INSTANCE_AARCH64(jit_aarch64_sve_512_common_dw_convolution_fwd_t) 
CPU_INSTANCE_AARCH64_ACL(arm_gemm_convolution_fwd_t) 
...
```
...will be perfectly fine for now, and I would not expect any conflicts with our planned implementation(#820).

Arm Compute Library does provide kernels for SVE targets, if Compute Library has been built for that target arch. At the moment our focus is on Neon targets, but Compute Library can, and we would anticipate will, provide SVE kernels. With the suggested ordering, this would have the advantage of Compute Library providing an SVE implementation for vector lengths not supported by the JITed methods.

In the longer term dispatch would, I expect, need to be based on empirical data, with heuristics to guide the selection - these could potentially be a function of the microarchitecture, problem size etc. I think this is consistent with @kawakami-k's assessment, but expanded to cover SVE as well as non-SVE implementations.